hummingbot/data_feed/candles_feed/candles_base.py

```diff
@@ -199,21 +298,88 @@ async def listen_for_subscriptions(self):
 
     async def _connected_websocket_assistant(self) -> WSAssistant:
         ws: WSAssistant = await self._api_factory.get_ws_assistant()
-        await ws.connect(ws_url=self.wss_url,
-                         ping_timeout=30)
+        await ws.connect(ws_url=self.wss_url, ping_timeout=30)
         return ws
 
     async def _subscribe_channels(self, ws: WSAssistant):
         """
         Subscribes to the candles events through the provided websocket connection.
         :param ws: the websocket assistant used to connect to the exchange
         """
+        try:
+            subscribe_candles_request: WSJSONRequest = WSJSONRequest(payload=self.ws_subscription_payload())
+            await ws.send(subscribe_candles_request)
+            self.logger().info("Subscribed to public klines...")
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            self.logger().error(
+                "Unexpected error occurred subscribing to public klines...",
+                exc_info=True
+            )
+            raise
+
+    def ws_subscription_payload(self):
+        """
+        This method returns the subscription payload for the websocket connection.
+        """
         raise NotImplementedError
 
     async def _process_websocket_messages(self, websocket_assistant: WSAssistant):
+        # TODO: Implement a way to handle ping pong messages
+        async for ws_response in websocket_assistant.iter_messages():
+            data = ws_response.data
+            parsed_message = self._parse_websocket_message(data)
+            # parsed messages may be ping or pong messages
+            if isinstance(parsed_message, WSJSONRequest):
```

Comment:
this is in case that the message received is a ping?

Reply:
yes, it can be ping or pong depending the exchange. the best way is to isolate this feature and manage it through mocking in the implementations.

i need to think the best way to improve this, probably will contribute it in another pr